core/pkg/server/handler.go

```diff
@@ -972,35 +966,48 @@ func (h *Handler) handleRequestJobInput(record *service.Record) {
 	h.fwdRecord(record)
 }
 
-func (h *Handler) handleSummary(record *service.Record, summary *service.SummaryRecord) {
-	if !h.settings.GetXSync().GetValue() {
-		// if sync is enabled, we don't need to do all this
-		runtime := int32(h.runTimer.Elapsed().Seconds())
+// handleSummary processes an update to the run summary.
+//
+// These records come from one of three sources:
+//   - Explicit updates made by using `run.summary[...] = ...`
+//   - Records from the transaction log when syncing
+//   - `updateRunTiming`
+func (h *Handler) handleSummary(
+	record *service.Record,
+	summary *service.SummaryRecord,
+) {
+	for _, update := range summary.Update {
+		err := h.runSummary.SetFromRecord(update)
+		if err != nil {
+			h.logger.CaptureError(
+				fmt.Errorf("handler: error processing summary: %v", err))
+		}
+	}
 
-		// update summary with runtime
-		summary.Update = append(summary.Update, &service.SummaryItem{
-			NestedKey: []string{"_wandb", "runtime"},
-			ValueJson: fmt.Sprintf("%d", runtime),
-		})
+	for _, remove := range summary.Remove {
+		h.runSummary.RemoveFromRecord(remove)
 	}
 
-	h.runSummary.ApplyChangeRecord(
-		summary,
-		func(err error) {
-			h.logger.CaptureError(
-				fmt.Errorf("error updating run summary: %v", err))
-		},
-	)
+	h.fwdRecord(record)
+}
 
-	h.fwdRecordWithControl(record,
-		func(control *service.Control) {
-			control.AlwaysSend = true
-			// do not write to the transaction log when syncing an offline run
-			if h.settings.GetXSync().GetValue() {
-				control.Local = true
-			}
+// updateRunTiming updates the `_wandb.runtime` summary metric which
+// tracks the total duration of the run.
+func (h *Handler) updateRunTiming() {
+	runtime := int(h.runTimer.Elapsed().Seconds())
+	record := &service.Record{
+		RecordType: &service.Record_Summary{
+			Summary: &service.SummaryRecord{
+				Update: []*service.SummaryItem{{
+					Key:       "_wandb.runtime",
+					ValueJson: strconv.Itoa(runtime),
+				}},
+			},
 		},
-	)
+		Control: &service.Control{Local: true},
```

Comment:
how will the transaction log have the final runtime value to sync for offline?

Reply:
oh, good catch!